File: apps/fsharp/src/MachineLearning/Chapter09/Outliers.fs
module MachineLearning.Chapter09.Outliers

open MachineLearning.Chapter02.IrisPreprocessing

/// 分位点。値を並べ、位置 (件数 - 1) × q の前後の値を線形補間する
let quantile (values: float list) (q: float) : float =
    let sorted = values |> List.sort |> List.toArray
    let position = float (sorted.Length - 1) * q
    let lower = sorted[int (floor position)]
    let upper = sorted[int (ceil position)]
    lower + (upper - lower) * (position - floor position)

/// 第 1 四分位点・第 3 四分位点から、四分位範囲（IQR）の k 倍より外側にある値を外れ値とする
let iqrOutliersWith (k: float) (values: float list) : bool list =
    let q1 = quantile values 0.25
    let q3 = quantile values 0.75
    let iqr = q3 - q1

    values |> List.map (fun value -> value < q1 - k * iqr || value > q3 + k * iqr)

/// k = 1.5 の四分位範囲による外れ値
let iqrOutliers: float list -> bool list = iqrOutliersWith 1.5

/// 訓練データの正解の外れ値の行を、訓練データから除く。テストデータには手を付けない
let removeTargetOutliers (split: TrainTestSplit<'X, float>) : TrainTestSplit<'X, float> =
    let kept =
        List.zip3 split.XTrain split.TTrain (iqrOutliers split.TTrain)
        |> List.filter (fun (_, _, outlier) -> not outlier)

    { split with
        XTrain = kept |> List.map (fun (x, _, _) -> x)
        TTrain = kept |> List.map (fun (_, t, _) -> t)
    }

File: apps/ruby/lib/getting_started_ml/chapter09/outliers.rb
# frozen_string_literal: true

module GettingStartedMl
  # 第 9 章の外れ値の検出。
  module Chapter09
    # 外れ値とみなす範囲の、四分位範囲に掛ける倍率。
    DEFAULT_K = 1.5

    module_function

    # 分位数を求める。位置が値の間にあれば前後の値から線形補間する
    # （pandas の quantile の既定と同じ）。
    def quantile(values, ratio)
      raise ArgumentError, "値が 1 件もありません" if values.empty?

      sorted = values.sort
      position = (sorted.size - 1) * ratio
      lower = sorted[position.floor]

      lower + ((sorted[position.ceil] - lower) * (position - position.floor))
    end

    # 四分位範囲（IQR）の factor 倍より外側にある値を true にした並びを返す。
    def iqr_outliers(values, factor = DEFAULT_K)
      q1 = quantile(values, 0.25)
      q3 = quantile(values, 0.75)
      range = (q1 - (factor * (q3 - q1)))..(q3 + (factor * (q3 - q1)))

      values.map { |value| !range.cover?(value) }
    end

    # 訓練データから、正解の値が外れ値の行を取り除く。テストデータはそのまま残す。
    def remove_target_outliers(split)
      outliers = iqr_outliers(split.t_train)
      kept = split.x_train.zip(split.t_train).reject.with_index { |_, index| outliers[index] }
      x_train, t_train = kept.transpose

      split.with(x_train: x_train || [], t_train: t_train || [])
    end
  end
end
